lib/pbench/test/unit/agent/test_config.py

```diff
@@ -34,3 +38,72 @@ def test_results_attr():
         "pbench"
         in PbenchAgentConfig(os.environ["_PBENCH_AGENT_CONFIG"]).results["user"]
     )
+
+
+# Monkeypatching different Classes and Methods to test Pbench-run value
+
+
+class MockPbenchAgentConfig:
+    """Mock class for PbenchAgentConfig"""
+
+    def __init__(self, cfg_name: str):
+        self.pbench_run = "/test/pbench-run"
+        self.pbench_log = "logger"
+        self.pbench_install_dir = Path("/test/install_dir")
+        self.ssh_opts = "/ssh_opts"
+        self.scp_opts = "/scp_opts"
+
+
+def mock_exist(path: Path):
+    """mock method for Path.exist"""
+    return True
+
+
+def mock_current_context():
+    """mock get_current_context from click package"""
+    None
+
+
+class MockContext:
+    """Mocking Context"""
+
+    config = "/test/pbench-config-path"
+
+
+class TestCommand(BaseCommand):
+    """Mock of class that extends BaseCommand"""
+
+    def __init__(self):
+        super(TestCommand, self).__init__(MockContext())
+
+    def execute(self):
+        pass
+
+
+def test_pbench_run_path(monkeypatch):
+    """Use pbench_run value from pbench-agent.cfg file"""
+    assert os.environ.get("pbench_run") is None
+    monkeypatch.setattr(PbenchAgentConfig, "__init__", MockPbenchAgentConfig.__init__)
+    monkeypatch.setattr(Path, "exists", mock_exist)
+    pbench_run_val = TestCommand().pbench_run
+    assert str(pbench_run_val) == "/test/pbench-run"
+
+
+def test_pbench_run_as_env(monkeypatch):
+    """Use pbench_run value, passed manually through
+    `pbench-run` environment variable"""
+    monkeypatch.setenv("pbench_run", "/test/pbench-run/environ")
+    monkeypatch.setattr(PbenchAgentConfig, "__init__", MockPbenchAgentConfig.__init__)
+    monkeypatch.setattr(Path, "exists", mock_exist)
+    pbench_run_val = TestCommand().pbench_run
+    assert str(pbench_run_val) == "/test/pbench-run/environ"
+
+
+def test_pbench_run_dir_existence(monkeypatch):
+    """Verify that pbench_list_tools fail when pbench_run is defined
+    to a directory that doesn't exist"""
+    monkeypatch.setattr(PbenchAgentConfig, "__init__", MockPbenchAgentConfig.__init__)
+    monkeypatch.setattr(click, "get_current_context", mock_current_context)
+    pbench_run = TestCommand().pbench_run
+    assert pbench_run
+    # assert pbench_run.startswith("[ERROR] pbench run directory does not exist")
```

Comment:
We should be mocking the `exists()` function to return `False` for this test (rather than counting on make a file system access and having it fail).

As for checking the contents of `stderr`, look at the Click [Testing](https://click.palletsprojects.com/en/7.x/api/#testing) support; [here's an example](https://github.com/riya-17/pbench/blob/0b3c2d4029aaccbbcb2b5fc3b855c48591551a25/lib/pbench/test/unit/agent/task/test_generate_token.py#L138) of how we use it.  (I think it will work with your mock-`TestCommand()`.)

Once the `exists()` call returns `False`, the error message will be sent to `stderr`, which the test runner will capture in the `Result` object, and you can reference that in an assertion, once you get control back after catching the exception from the `exit()` mock.

(Let me know if you have any questions or need any more explanation or help with this.)

Reply:
Thank you this is what I was looking for

Reply:
Thinking about this this morning, if you use the Click test runner, then you probably won't need to mock `exit()`.  So, I would try using the test runner first and defer mocking `exit()` until it proves to be needed.